### DISCUSSION TITLE
twinbasic as a replacement for the VBA IDE

### DISCUSSION THREAD

**Discussion by gregse99** [twinbasic/twinbasic#631] [Q&A]
So let me preamble this with a disclaimer that I am not a professional developer but rather an advanced user of Excel who programs at an intermediate level in VBA.  So my question may reflect my ignorance on advanced programming subjects.

My question is in response to Wayne's objective of eventually replacing the VBA IDE with twinbasic.  This sounds like a great goal but there is one library that would be directly affected by this: the VBA object library that lets you manipulate the VBA IDE itself.  I happen to use that library extensively to programatically automate my coding process (e.g., creating new modules with method templates).  My question is this: is there a similar way of automating such things in VS Code (perhaps v. an API)?  Technically speaking, that would be the one thing I can think of that would make 100% backwards-compatability impossible for VBA at least (I never used VB6), right?  Hopefully this is not a dumb question.

**Comment by retailcoder** [comment]
That would be the (eventual) extensibility API 😉

Side note, the VBIDE Extensibility API is objectively a rather poor API, giving you exactly zero insights about the code or its structure: for the most part it treats code as, essentially, strings (omitting the hidden parts!).. which is exactly why Rubberduck had to implement its own parser+resolver!

**Reply by mwolfe02** [reply]
I have high hopes (and expectations!) for the eventual tB extensibility API.  I can't wait to see what sort of tools you and the rest of the Rubberduck VBA community come up with when you're not spending more than half your (development- and execution-)time re-parsing the language itself.

**Comment by WaynePhillipsEA** [comment]
Indeed, we hope that the analysis extension API #51 will largely replace the need for the VBIDE library, by providing much deeper access to the guts of the compiler.  However, for backwards compatibility we will very likely end up implementing a matching version of the VBIDE object model as well.

**Reply by mburns08109** [reply]
Wouldn't it make the most sense to create the NEW API approach first, and then just use a(n) (dumbed-down) interface as a pretty-much direct replacement for the older VBIDE Extensibility 5.5 reference?

**Reply by WaynePhillipsEA** [reply]
Yes, that is the plan.  Sorry if I wasn't clear on that.

**Comment by mburns08109** [comment]
Doesn't this beg an implementation question though?
If I understand the thinking here, we're talking about making VSSCode + tB an effective replacement for VBA, right?
...but how would that actually work in practice?
vbx doesn't compile to a .exe, but to pcode that I'm guessing the VBA runtime then reads and executes alongside the ms-office apps .exe, right?
How could it work for a tB-compiled .exe/.dll to work in direct place of that VBA/p-code in practice?
Register the tB as an activeX .dll add-in? (Once again, there are some serious problems with that in enterprise configurations, as default enterprise security profiles generally lock those down.)

**Reply by WaynePhillipsEA** [reply]
Ahh, well I might have a few tricks up my sleeve.... you'll have to wait and see 😜

**Reply by mansellan** [reply]
Heh, I think I may have an inkling here, but... _shh... spoilers!_
